posts/Yoann/2019-11-08-hibernate-search-quarkus.adoc

```diff
@@ -0,0 +1,1428 @@
+
+[source]
+----
+curl -X GET 'http://localhost:8080/client/search/?terms=laboratory'
+
+[
+]
+----
+
+We didn't find any match. That's annoying, because one of our clients is called "Aperture Science Laboratories".
+It's not an exact match, but still, users of our application
+would expect that client to turn up when they type "laboratory" (singular).
+
+Full-text search allows us to tackle that kind of "non-exact" matches thanks to what is called
+link:{hsearchDocUrl}#concepts-analysis[analysis].
+Simply put, analysis is the process of transforming text, both during indexing (transforming the indexed text)
+and searching (transforming the terms of the search query).
+It is used to extract tokens (words) from text, but also to normalize these words.
+For example, a correctly configured analyzer will transform "Laboratories" into "laboratory",
+so that when we search for the word "laboratory", the name "Aperture Science Laboratories" will match.
+
+In order to take advantage of analysis, we need to configure analyzers.
+Hibernate Search provides APIs to easily configure analyzers,
+and will automatically push the analyzer definitions to Elasticsearch when it creates the indexes.
+
+We just need to implement an analyzer configurer:
+
+[source,JAVA]
+----
+package org.hibernate.demos.quarkus.search;
+
+import org.hibernate.search.backend.elasticsearch.analysis.ElasticsearchAnalysisConfigurationContext;
+import org.hibernate.search.backend.elasticsearch.analysis.ElasticsearchAnalysisConfigurer;
+
+public class ClientElasticsearchAnalysisConfigurer implements ElasticsearchAnalysisConfigurer {
+	@Override
+	public void configure(ElasticsearchAnalysisConfigurationContext context) {
+		context.analyzer( "english" ).custom() // <1>
+				.tokenizer( "standard" ) // <2>
+				.tokenFilters( "lowercase", "stemmer_english", "asciifolding" ); // <3>
+		context.tokenFilter( "stemmer_english" ) // <4>
+				.type( "stemmer" )
+				.param( "language", "english" );
+	}
+}
+----
+<1> Declare a custom analyzer named `english`.
+<2> Set the tokenizer to `standard`, i.e. require that the analyzer generates words by splitting text on spaces, tabs, punctuation, etc.
+<3> Apply three token filters to transform the extracted words: `lowercase` which turns the words to lowercase,
+`stemmer_english` which is a custom filter (see below), `asciifolding` which replaces accented characters with they ascii counterpart
+(`déjà-vu` => `deja-vu`).
+<4> Declare a custom token filter named `stemmer_english`.
+This token filter is a stemmer, meaning it will normalize the end of words (`laboratories` => `laboratory`),
+and we configure it to handle the English language.
+
+Then, we need to tell Hibernate Search to use our configurer by setting a configuration property in `application.properties`:
+
+[source]
+----
+quarkus.hibernate-search.elasticsearch.analysis.configurer=org.hibernate.demos.quarkus.search.ClientElasticsearchAnalysisConfigurer
+----
```

Comment:
See how cool it would be to just have an annotation ;)

Reply:
Yep, it's called `@Named` but for some obscure reason it won't work in Quarkus :)